huntsman/camera/pyro.py

```diff
@@ -0,0 +1,498 @@
+
+        if focus_step is not None:
+            focus_step = (int(step) for step in focus_step)
+
+        if keep_files is not None:
+            keep_files = bool(keep_files)
+
+        if take_dark is not None:
+            take_dark = bool(take_dark)
+
+        if thumbnail_size is not None:
+            thumbnail_size = int(thumbnail_size)
+
+        merit_function = str(merit_function)
+        merit_function_kwargs = dict(merit_function_kwargs)
+
+        if mask_dilations is not None:
+            mask_dilations = int(mask_dilations)
+
+        if coarse is not None:
+            coarse = bool(coarse)
+
+        if make_plots is not None:
+            make_plots = bool(make_plots)
+
+        if isinstance(timeout, u.Quantity):
+            timeout = timeout.to(u.second).value
+        if timeout is not None:
+            timeout = float(timeout)
+
+        # Compile aruments into a dictionary
+        autofocus_kwargs = {'seconds': seconds,
+                            'focus_range': focus_range,
+                            'focus_step': focus_step,
+                            'keep_files': keep_files,
+                            'take_dark': take_dark,
+                            'thumbnail_size': thumbnail_size,
+                            'merit_function': merit_function,
+                            'merit_function_kwargs': merit_function_kwargs,
+                            'mask_dilations': mask_dilations,
+                            'coarse': coarse,
+                            'make_plots': make_plots}
+        autofocus_kwargs.update(kwargs)
+
+        focus_dir = os.path.join(os.path.abspath(self.config['directories']['images']), 'focus/')
+
+        # Make sure proxy is in async mode
+        Pyro4.asyncproxy(self._proxy, asynchronous=True)
+
+        # Start autofocus
+        autofocus_result = {}
+        self.logger.debug('Starting autofocus on {}'.format(self.name))
+        # Remote method call to start the autofocus
+        autofocus_result = self._proxy.autofocus(*args, **autofocus_kwargs)
+        # Tag the file transfer on the end.
+        autofocus_result = autofocus_result.then(self._file_transfer, focus_dir)
+        # Tag empty directory cleanup on the end & keep future result to check for completion
+        autofocus_result = autofocus_result.then(self._clean_directories)
+
+        # Start a thread that will set an event once autofocus has completed
+        autofocus_event = Event()
+        autofocus_thread = Thread(target=self._async_wait,
+                                  args=(autofocus_result, 'autofocus', autofocus_event, timeout))
+        autofocus_thread.start()
+
+        if blocking:
+            autofocus_event.wait()
+
+        return autofocus_event
+
+# Private Methods
+
+    def _clean_directories(self, source):
+        """
+        Clean up empty directories left behind by rsysc.
+        """
+        user_at_host, path = source.split(':')
```

Comment:
There's some weirdness with the path going on. Maybe an example of what the `source` input is here.

Reply:
Ah, yes, the directory cleanup follows the file transfer in the call chain, which just passes on the same `source` argument that the file transfer uses, which is in `user@host:/directory/subdirectory/filename` format ready for rysnc.